internal/trie/node/header.go

```diff
@@ -4,44 +4,148 @@
+	// Partial key length does not fit in header byte only
+	buffer[0] |= partialKeyLengthMask
+	partialKeyLength -= int(partialKeyLengthMask)
+	_, err = writer.Write(buffer)
 	if err != nil {
 		return err
 	}
 
-	err = encodeKeyLength(len(node.Key), writer)
-	return err
+	for {
+		buffer[0] = 255
+		if partialKeyLength < 255 {
+			buffer[0] = byte(partialKeyLength)
+		}
+
+		_, err = writer.Write(buffer)
+		if err != nil {
+			return err
+		}
+
+		partialKeyLength -= int(buffer[0])
+
+		if buffer[0] < 255 {
+			break
+		}
+	}
+
+	return nil
+}
+
+var (
+	ErrPartialKeyTooBig = errors.New("partial key length cannot be larger than 2^16")
+)
+
+func decodeHeader(reader io.Reader) (variant byte,
+	partialKeyLength uint16, err error) {
+	buffer := make([]byte, 1)
+	_, err = reader.Read(buffer)
+	if err != nil {
+		return 0, 0, fmt.Errorf("cannot read header byte: %w", err)
+	}
+
+	variant, partialKeyLengthHeader, partialKeyLengthHeaderMask,
+		err := decodeHeaderByte(buffer[0])
+	if err != nil {
+		return 0, 0, fmt.Errorf("cannot decode header byte: %w", err)
+	}
+
+	partialKeyLength = uint16(partialKeyLengthHeader)
+	if partialKeyLengthHeader < partialKeyLengthHeaderMask {
+		// partial key length is contained in the first byte.
+		return variant, partialKeyLength, nil
+	}
+
+	// the partial key length header byte is equal to its maximum
+	// possible value; this means the partial key length is greater
+	// than this (0 to 2^6 - 1 = 63) maximum value, and we need to
+	// accumulate the next bytes from the reader to get the full
+	// partial key length.
+	// Specification: https://spec.polkadot.network/#defn-node-header
+	var previousKeyLength uint16 // used to track an eventual overflow
+	for {
+		_, err = reader.Read(buffer)
+		if err != nil {
+			return 0, 0, fmt.Errorf("cannot read key length: %w", err)
+		}
+
+		previousKeyLength = partialKeyLength
+		partialKeyLength += uint16(buffer[0])
+
+		if partialKeyLength < previousKeyLength {
+			// the partial key can have a length up to 65535 which is the
+			// maximum uint16 value; therefore if we overflowed, we went over
+			// this maximum.
+			overflowed := maxPartialKeyLength - previousKeyLength + partialKeyLength
+			return 0, 0, fmt.Errorf("%w: overflowed by %d", ErrPartialKeyTooBig, overflowed)
+		}
+
+		if buffer[0] < 255 {
+			// the end of the partial key length has been reached.
+			return variant, partialKeyLength, nil
+		}
+	}
+}
+
+var ErrVariantUnknown = errors.New("node variant is unknown")
+
+func decodeHeaderByte(header byte) (variantBits,
+	partialKeyLengthHeader, partialKeyLengthHeaderMask byte, err error) {
+	// variants is a slice of all variants sorted in ascending
+	// order by the number of bits each variant mask occupy
+	// in the header byte.
+	// See https://spec.polkadot.network/#defn-node-header
+	variants := []variant{
+		leafVariant,            // mask 1100_0000
+		branchVariant,          // mask 1100_0000
+		branchWithValueVariant, // mask 1100_0000
+	}
```

Comment:
could this be moved to a package level `var`?

Reply:
I don't mutation should be an issue, it should be pretty well understood we shouldn't be mutating `variants`.  I would assume `decodeHeader` is called a lot? We should be only allocating once if possible.

Reply:
> I would assume decodeHeader is called a lot? We should be only allocating once if possible.

Counter-intuitively, it really does not matter performance wise, it's the exact same speed/allocation. I pushed https://github.com/ChainSafe/gossamer/pull/2530/commits/374a29b6759981bcf91d56ae84bfda12cf158a65 with a comment + benchmark to clarify for future readers.

As a consequence, there is no need to make it global which would make it harder to read and pollute the Go package API a bit more.